### PR TITLE
 TECH-7631 admin.py clean-up and make more DRY try 2

### PR DESCRIPTION
**Pull request by MiriamSexton**
[TECH-7631 cleanup admin.py and make it more DRY](https://industrydive.atlassian.net/browse/TECH-7631) -- I also fixed the bug where fieldsets were not appearing on the add page, and changed the privacy field to readonly in many cases. 

### Non-technical Context
Removed a workaround where you could change a semiprivate person to public and therefore expose their details. Cleaned up add page.

### Technical Context
There was a lot of duplicated code, and manipulating the readonly fields in particular causes problems. I pulled things out into the `get_fields`, `get_fieldsets` and `get_readonly_fields` hooks that django wants you to use.

### Manual Testing Instructions

Have 2 normal users. Log in as 1 of them and create 3 sources, one of each privacy level. The add page should be organized. Now view each source, and edit as well. You should see all the data, and everything except the default readonly fields should be editable.

Log in as the 2nd person. You should only see 2 of the suorces. View both. The appropriate fields should be hidden. Edit both. You should not be able to change the hidden fields or the privacy level of the semi private source.  

Create a semiprivate interaction as user1. You should be able to edit it. As user2, view it (you should not see the notes) and edit it (you should not be able to edit the privacy level OR the notes themselves).

#### Developer

- [x] Ensure code complies with the [style guide](https://industrydive.atlassian.net/wiki/spaces/TECH/pages/639500407/Tech+Team+s+Style+Guides)
- <s>Devise edge cases to test the bug fix or feature being merged</s> DECIDED NOT TO IN CONSULTATION WITH @greglinch 
- [x] Make sure the code is clean and understandable
- [x] Ensure code is not overly inefficient
- [x] Ensure documentation is understandable and accurate, including:
    - [x] Code comments and doc strings
    - [N/A] Technical documentation

#### Reviewer

- [x] Ensure code complies with the [style guide](https://industrydive.atlassian.net/wiki/spaces/TECH/pages/639500407/Tech+Team+s+Style+Guides)
    - If there are few or minor issues, add inline GitHub comments
    - If there are many issues or it seems like the developer did not follow the style guide, leave a comment asking them to do so
- [x] Devise edge cases to test the bug fix or feature being merged
- [x] Make sure the code is clean and understandable
    - [x] Ask for inline comments on unclear sections
- [x] Ensure code is not overly inefficient
- [x] Ensure documentation is understandable and accurate, including:
    - [x] Code comments and doc strings
    - [na] Technical documentation

### CODE REVIEW

File: sources/admin.py [hunk 1]
```diff
@@ -227,7 +227,7 @@ class PersonAdmin(admin.ModelAdmin):
     list_filter = [IndustryFilter, ExpertiseFilter, OrganizationFilter, 'timezone', 'city', 'state', 'privacy_level']
     search_fields = ['city', 'country', 'email_address', 'expertise__name', 'first_name', 'language', 'name', 'notes', 'organization', 'state', 'title', 'type_of_expert', 'twitter', 'website']
     filter_horizontal = ['expertise', 'industries', 'organization']
-    readonly_fields = ['entry_method', 'entry_type', 'created_by']
+    readonly_fields = ['entry_method', 'entry_type', 'created_by', 'updated']
```
Comment (MiriamSexton) [inline sources/admin.py:262]: changed this after speaking with @greglinch

Comment (greglinch) [reply sources/admin.py:262]: 👍

**Comment by greglinch** [comment]
@MiriamSexton This is a big improvement, thanks so much! One remaining is that I wasn't able to edit the `privacy_level` even if I was the user who created the `Person` object. `Interaction` object privacy editing worked as expected.

**Comment by MiriamSexton** [comment]
@greglinch hmm, interesting. Thoguht I checked that but I might have missed it after final changes. I'll take a look Monday

File: sources/admin.py [hunk 2]
```diff
@@ -227,7 +227,7 @@ class PersonAdmin(admin.ModelAdmin):
     list_filter = [IndustryFilter, ExpertiseFilter, OrganizationFilter, 'timezone', 'city', 'state', 'privacy_level']
```
Comment (greglinch) [inline sources/admin.py]: I just noticed that after changing `timezone` to the `pytz` list, we have a million options in the list view. Could we include the following quick update with this PR?
```suggestion
    list_filter = [IndustryFilter, ExpertiseFilter, OrganizationFilter, 'city', 'state', 'privacy_level']
```

File: sources/admin.py [hunk 3]
```diff
@@ -384,12 +398,45 @@ def get_queryset(self, request):
...
+
+        hide_contact_data = self._determine_whether_to_hide_contact_data(request, obj)
+
+        if 'edit' not in request.GET:
+            # if we're not editing, all fields should be readonly
+            # use _return_fieldsets to get the correct fields for this permission level
+            current_fieldsets = self._return_fieldsets(hide_contact_data=hide_contact_data)
+            return flatten_fieldsets(current_fieldsets)
+        elif hide_contact_data:
+            # if we're editing,, and data is hidden from the user, we want the contact data to be readonly
```
Comment (greglinch) [inline sources/admin.py]:
```suggestion
            # If we're editing, and data is hidden from the user, we want the contact data to be readonly
```

File: sources/admin.py [hunk 4]
```diff
@@ -84,23 +84,52 @@ class InteractionAdmin(admin.ModelAdmin):
...
-            self.readonly_fields = ['created_by']
-        return self.changeform_view(request, object_id, form_url, extra_context)
+            return self._fields_always_readonly   
 
 
+    def get_fields(self, request, obj=None):
+        """
+        If the user doesn't have permission to view the notes, display the replacement notes field instead.
+
+        Use Django's built in hook for accessing fields. Manipulating self.fields directly can lead to problems.
```
Comment (greglinch) [inline sources/admin.py]:
```suggestion
        Use Django's built in hook for accessing fields. NOTE: Manipulating self.fields directly can lead to problems.
```

File: sources/admin.py [hunk 5]
```diff
@@ -84,23 +84,52 @@ class InteractionAdmin(admin.ModelAdmin):
...
+
+    def get_readonly_fields(self, request, obj=None):
         """
-        If the privacy level is semi-private/searchable and the logged
-        in user is not the one who created the interaction, then remove
-        notes field for them
+        If the user doesn't have permission to see the notes, both the notes display and the
+        privacy level are added to the default readonly fields.
+
+        Use Django's built in hook for accessing readonly fields. Manipulating self.readonly_fields
```
Comment (greglinch) [inline sources/admin.py]:
```suggestion
        Use Django's built in hook for accessing readonly fields. NOTE: Manipulating self.readonly_fields
```

File: sources/admin.py [hunk 6]
```diff
@@ -252,123 +281,108 @@ def phone_number_secondary_semiprivate_display(self, obj):
...
-                'email_address_semiprivate_display',
-                'phone_number_primary_semiprivate_display',
-                'phone_number_secondary_semiprivate_display',
-                'updated',
-            ]
+        if not obj:
+            # If creating a new `Person`, give all permissions
+            return False
+        elif obj.created_by == request.user:
+            # users can always see their own `Person`s
```
Comment (greglinch) [inline sources/admin.py]:
```suggestion
            # Users can always see their own `Person`s
```

File: sources/admin.py [hunk 7]
```diff
@@ -252,123 +281,108 @@ def phone_number_secondary_semiprivate_display(self, obj):
...
+            return False
+        elif obj.privacy_level in ['searchable', 'private_individual']:
+            # if the `Person` is at all private, and the user didn't create the
```
Comment (greglinch) [inline sources/admin.py]: After being a devoted "all lowercase to start comments person," I'm starting to come around to capitalizing — especially in this case so it doesn't initially look like a commented out conditional statement.

```suggestion
            # If the `Person` is at all private, and the user didn't create the
```

File: sources/admin.py [hunk 8]
```diff
@@ -252,123 +281,108 @@ def phone_number_secondary_semiprivate_display(self, obj):
...
-                }),
-            )
-            self.readonly_fields = [
-                'created_by',
-                'entry_method',
-                'entry_type',
-                'updated',
-            ]
-        return self.changeform_view(request, object_id, form_url, extra_context)
+            # only reach here for public `Person`s created by a different user
```
Comment (greglinch) [inline sources/admin.py]:
```suggestion
            # Only reach here for public `Person`s created by a different user
```

File: sources/admin.py [hunk 9]
```diff
@@ -384,12 +398,45 @@ def get_queryset(self, request):
...
 
 
+    def get_fieldsets(self, request, obj=None):
+        """
+        Use Django's built in hook for accessing the fieldsets. Manipulating self.fieldsets directly
+        leads to problems.
+        """
+
+        if obj and obj.created_by != request.user and obj.privacy_level == 'private_individual':
+            # cover the one weird edge case not covered by `_determine_whether_to_hide_contact_data`
```
Comment (greglinch) [inline sources/admin.py]:
```suggestion
            # Cover the one weird edge case not covered by `_determine_whether_to_hide_contact_data`
```

File: sources/admin.py [hunk 10]
```diff
@@ -384,12 +398,45 @@ def get_queryset(self, request):
...
+            # this is if the user is trying to view a person they're not even allowed to know exists
```
Comment (greglinch) [inline sources/admin.py]:
```suggestion
            # This is if the user is trying to view a person they're not even allowed to know exists
```

File: sources/admin.py [hunk 11]
```diff
@@ -384,12 +398,45 @@ def get_queryset(self, request):
...
+            return [(None, {'fields': []})]
+        else:
+            # construct the correct fieldsets based on permissions
```
Comment (greglinch) [inline sources/admin.py]:
```suggestion
            # Construct the correct fieldsets based on permissions
```

File: sources/admin.py [hunk 12]
```diff
@@ -384,12 +398,45 @@ def get_queryset(self, request):
...
+    def get_fieldsets(self, request, obj=None):
+        """
+        Use Django's built in hook for accessing the fieldsets. Manipulating self.fieldsets directly
+        leads to problems.
+        """
+
+        if obj and obj.created_by != request.user and obj.privacy_level == 'private_individual':
+            # cover the one weird edge case not covered by `_determine_whether_to_hide_contact_data`
+            # this is if the user is trying to view a person they're not even allowed to know exists
+            return [(None, {'fields': []})]
```
Comment (greglinch) [inline sources/admin.py:418]: Smart backup!

File: sources/admin.py [hunk 13]
```diff
@@ -384,12 +398,45 @@ def get_queryset(self, request):
...
     def get_readonly_fields(self, request, obj=None):
-        if self.fieldsets and 'edit' not in request.GET:
-            return flatten_fieldsets(self.fieldsets)
-        elif 'add' in request.GET:
+        """
+        Use Django's built in hook for accessing readonly fields. Manipulating self.readonly_fields
+        directly leads to problems.
+        """
+        if not obj:
+            # if creating the obj, use only default readonly fields
```
Comment (greglinch) [inline sources/admin.py]:
```suggestion
            # If creating the obj, use only default readonly fields
```

File: sources/admin.py [hunk 14]
```diff
@@ -384,12 +398,45 @@ def get_queryset(self, request):
...
+            # use _return_fieldsets to get the correct fields for this permission level
+            current_fieldsets = self._return_fieldsets(hide_contact_data=hide_contact_data)
+            return flatten_fieldsets(current_fieldsets)
+        elif hide_contact_data:
+            # if we're editing,, and data is hidden from the user, we want the contact data to be readonly
+            # we also want the privacy level field to be readonly
+            contact_fields_to_add = self._get_correct_contact_field_names(hide_contact_data=hide_contact_data)
+            return self.readonly_fields + ['privacy_level'] + contact_fields_to_add
         else:
+            # if we're editing AND the user has permissions, use the default fields
```
Comment (greglinch) [inline sources/admin.py]:
```suggestion
            # If we're editing AND the user has permissions, use the default fields
```

File: sources/admin.py [hunk 15]
```diff
@@ -384,12 +398,45 @@ def get_queryset(self, request):
...
+        hide_contact_data = self._determine_whether_to_hide_contact_data(request, obj)
+
+        if 'edit' not in request.GET:
+            # if we're not editing, all fields should be readonly
+            # use _return_fieldsets to get the correct fields for this permission level
+            current_fieldsets = self._return_fieldsets(hide_contact_data=hide_contact_data)
+            return flatten_fieldsets(current_fieldsets)
+        elif hide_contact_data:
+            # if we're editing,, and data is hidden from the user, we want the contact data to be readonly
+            # we also want the privacy level field to be readonly
```
Comment (greglinch) [inline sources/admin.py]:
```suggestion
            # We also want the privacy level field to be readonly
```

File: sources/admin.py [hunk 16]
```diff
@@ -384,12 +398,45 @@ def get_queryset(self, request):
...
+        directly leads to problems.
+        """
+        if not obj:
+            # if creating the obj, use only default readonly fields
             return self.readonly_fields
+
+        hide_contact_data = self._determine_whether_to_hide_contact_data(request, obj)
+
+        if 'edit' not in request.GET:
+            # if we're not editing, all fields should be readonly
```
Comment (greglinch) [inline sources/admin.py]:
```suggestion
            # If we're not editing, all fields should be readonly
```

File: sources/admin.py [hunk 17]
```diff
@@ -384,12 +398,45 @@ def get_queryset(self, request):
...
+            # use _return_fieldsets to get the correct fields for this permission level
```
Comment (greglinch) [inline sources/admin.py]:
```suggestion
            # Use _return_fieldsets to get the correct fields for this permission level
```

**Review by greglinch — APPROVED** [review]
@MiriamSexton Looks good, thanks so much for all your work on this! This is a great new approach.

The suggestions are all minor updates to the comments.

File: sources/admin.py (hunk 12 above)
Comment (MiriamSexton) [reply sources/admin.py:418]: Or just paranoia :)

Comment (MiriamSexton) [reply sources/admin.py:418]: Did you test this edge case?
